README.md
Node EasyMySQL
==============

## Purpose

A small collection of simple functions that make using MySQL with node.js easier. 
Uses the [node-mysql](https://github.com/felixge/node-mysql) library.

Developed by [MOG.com](http://mog.com)

## Installation

```
npm install easy-mysql
```

## Overview

EasyMySQL allows you to handle MySQL connection pooling, acquiring connections, and releasing them with ease.
For example, where you might do this (using node-pool):

    var get_widget = function(id, cb) {
        pool.acquire(function (err, client) {
            if (err) {
                pool.release(client);
                cb(err, null);
            } else {
                var sql = 'select * from widgets where id = ?';
                client.query(sql, [id], function (err, results) {
                    pool.release(client);
                    if (err || results && results.length === 0) {
                        cb(err, null);
                    } else {
                        cb(null, results[0]);
                    }
                });
            }
        });
    };

...you can do this instead with EasyMySQL:

    var get_widget = function(id, cb) {
        var sql = 'select * from widgets where id = ?';
        easy_mysql.get_one(sql, [id], cb);
    };

## Usage

    // require the module
    var EasyMySQL = require('easy-mysql');

### Connection Modes

You can connect three different ways with EasyMySQL:

  1. Direct: Directly establish a single connection for each query.
  This is probably not a good idea for production code, but may be fine for code
  where you don't want to set up a pool, such as in unit tests.

  2. Custom Pool: Pass in your own pool object.  It must have functions named
  'acquire' and 'release'.

  3. Built-in Pool: Use the built-in pool, which uses [node-pool](https://github.com/coopernurse/node-pool)


#### Direct Connect Example

    var settings = {
        user     : 'myuser',
        password : 'mypass',
        database : 'mydb'
    };
 
    var easy_mysql = EasyMySQL.connect(settings);

#### Custom Pool Example

    var my_pool = /* create your own pool here */;
    var easy_mysql = EasyMySQL.connect_with_pool(pool);

#### Built-in Pool Example

    var settings = {
        user      : 'myuser',
        password  : 'mypass',
        database  : 'mydb',
        pool_size : 50
    };

    var easy_mysql = EasyMySQL.connect_with_easy_pool(settings);

#### Logging

You can specify a logger to log error events (more event types may be added later).

    var settings = {
        user     : 'myuser',
        password : 'mypass',
        database : 'mydb',
        logging: {
            logger: my_logger,
            events: {
                error: {level: 'warn'}
            }
        }
    };

You must pass in your own logger and it must have a function with the same
name as the `level` you specify.


### Instance methods

#### get_one / getOne

Returns only one result, and if no results are found, returns null.

    var sql = 'select * from widgets where id = ?';
    easy_mysql.get_one(sql, [123], function (err, result) {
        cb(err, result);
    });


#### get_all / getAll

Returns an array of results, and if no results are found, returns an empty array.

    var sql = 'select * from widgets where id > ?';
    easy_mysql.get_all(sql, [123], function (err, results) {
        cb(err, results);
    });

#### execute

Executes an arbitrary SQL query and returns the results from node-mysql.

    var sql = 'update widgets set foo = 'bar' where id = ?';
    easy_mysql.execute(sql, [123], function (err, results) {
        cb(err, results);
    });

## Tests and Docs

To run tests and generate docs, first run:

    npm install -d

Run the tests:

    make test

Generate JSDocs:

    make doc


## Contribute

If you would like to contribute to the project, please fork it and send us a pull request.  Please add tests
for any new features or bug fixes.  Also run ``make lint`` before submitting the pull request.


## License

node-easy-mysql is licensed under the MIT license.


History.md
- 0.0.1
  Initial release.

- 0.5.0
  Converted tests from nodeunit to mocha.

- 0.6.0
  Added support for test coverage (via mocha). Requires jscoverage. Run with ``make test-cov``.

- 0.6.1
  Added support for multiple pools in easy_pool. Fixes https://github.com/Mog-Inc/easy-mysql/issues/1.

- 0.6.2
  Using ENV var EASY_MYSQL_JSCOV to avoid collisions if others use the JSCOV ENV var.

- 0.6.3
  Adding alias methods all(), one(), and query(). Bit more test coverage.

- 0.6.4
  Adding compatibility with node-mysql 0.9.1 (to be deprecated) and 2.0.0-alpha.

- 0.6.5
  Cleanup, bumping generic-pool version to 2.0.2.

- 0.6.6
  Fixed intermittent stress test failure (solution: use InnoDB, not MyISAM).

- 1.0.0
  Version bump.  This has been stable for a while now.

- 1.0.1
  Adding ability to add a logger to log errors returned from MySQL.

- 1.0.2
  Making logging work when using EasyMySQL.connect_with_pool().

- 1.0.3
  Switching to Istanbul for test coverage.


test/run.js
#!/usr/bin/env node
process.env.NODE_ENV = 'test';
require('../index');

var Mocha = require('mocha');
var optimist = require('optimist');
var walk_dir = require('./support/walk_dir');

var argv = optimist
    .usage("Usage: $0 -T [types] --reporter [reporter] --timeout [timeout]")
    .default({types: 'unit,component', reporter: 'spec', timeout: 6000})
    .describe('types', 'The types of tests to run, separated by commas. E.g., unit,functional,acceptance')
    .describe('reporter', 'The mocha test reporter to use.')
    .describe('timeout', 'The mocha timeout to use per test (ms).')
    .boolean('bail')
    .describe('bail', 'Bail on first failure')
    .boolean('verbose')
    .describe('verbose', 'Verbose output')
    .boolean('help')
    .alias('types', 'T')
    .alias('timeout', 't')
    .alias('reporter', 'R')
    .alias('bail', 'b')
    .alias('help', 'h')
    .argv;

var mocha = new Mocha({timeout: argv.timeout, reporter: argv.reporter, ui: 'bdd'});
mocha.checkLeaks();
if (argv.bail) {
    mocha.bail();
}

var valid_test_types = ['unit', 'functional', 'acceptance', 'integration'];
var requested_types = argv.types.split(',');
var types_to_use = [];

valid_test_types.forEach(function (valid_test_type) {
    if (requested_types.indexOf(valid_test_type) !== -1) {
        types_to_use.push(valid_test_type);
    }
});

if (argv.help || types_to_use.length === 0) {
    console.log('\n' + optimist.help());
    process.exit();
}

var is_valid_file = function (file) {
    for (var i = 0; i < types_to_use.length; i++) {
        var test_type = types_to_use[i];
        var ext = test_type + ".js";

        if (file.indexOf(ext) !== -1) {
            return true;
        }
    }

    return false;
};

var test_files = [];

function run(cb) {
    walk_dir.walk('test', is_valid_file, function (err, files) {
        if (err) { return cb(err); }

        files.forEach(function (file) {
            test_files.push(file);
            mocha.addFile(file);
        });

        cb();
    });
}

run(function (err) {
    if (err) { throw err; }

    if (argv.verbose) {
        console.log("\nTest files:");
        test_files.forEach(function (file) {
            console.log(file + " \\");
        });
    }
    mocha.run(function (failures) {
        process.exit(failures);
    });
});


test/clone.js
function clone(object) {
    var ret = {};
    Object.keys(object).forEach(function (val) {
        ret[val] = object[val];
    });
    return ret;
}

module.exports = clone;


test/common.js
var mysql = require('mysql');
var assert = require('assert');

var lib_dir = 'lib';
if (process.env.EASY_MYSQL_JSCOV) {
    lib_dir = 'lib-cov';
}

var EasyMySQL = require('../' + lib_dir + '/easy_mysql');
var EasyClient = require('../' + lib_dir + '/easy_client');
var easy_pool = require('../' + lib_dir + '/easy_pool');

var conn_method = 'createConnection';
if (mysql.hasOwnProperty('createClient')) {
    conn_method = 'createClient';
}

function clone(object) {
    var ret = {};
    Object.keys(object).forEach(function (val) {
        ret[val] = object[val];
    });
    return ret;
}

function random_string() {
    return 'str' + Math.floor(Math.random() * 10000000000);
}

try {
    var settings = require('./settings');
} catch (e) {
    console.log('Skipping. See test/settings.example.js for more information.');
    process.exit(0);
}

var db = settings.db1.database;

var table_sql = "create table widgets( " +
                "id int auto_increment primary key, " +
                "name varchar(25)) Engine=InnoDB ";

function setup_db(cb) {
    var _settings = clone(settings.db1);
    delete _settings.database;
    var client;
    try {
        client = mysql[conn_method](_settings);
    } catch (e) {
        client = new mysql.Client(_settings);
        client.connect();
    }

    client.query('CREATE DATABASE ' + db, function (err) {
        if (!(err && (err.number == mysql.ERROR_DB_CREATE_EXISTS || err.number == 1007))) {
            throw err;
        }
    });
    client.query('USE ' + db);
    client.query('drop table if exists widgets');
    client.query(table_sql, function (err) {
        client.end();
        assert.ifError(err);
        cb(null, true);
    });
}

function setup_db2(cb) {
    var _settings = clone(settings.db2);
    delete _settings.database;
    var client;
    try {
        client = mysql[conn_method](_settings);
    } catch (e) {
        client = new mysql.Client(_settings);
        client.connect();
    }

    client.query('CREATE DATABASE ' + settings.db2.database, function (err) {
        client.end();
        if (!(err && (err.number == mysql.ERROR_DB_CREATE_EXISTS || err.number == 1007))) {
            throw err;
        }
        cb(null);
    });
}

var fake_logger = {
    debug: function () {},
    info: function () {},
    warn: function () {},
    error: function () {}
};

exports.setup_db      = setup_db;
exports.setup_db2     = setup_db2;
exports.database      = 'easy_mysql_test';
exports.settings      = settings;
exports.clone         = clone;
exports.random_string = random_string;
exports.EasyMySQL     = EasyMySQL;
exports.EasyClient    = EasyClient;
exports.easy_pool     = easy_pool;
exports.fake_logger   = fake_logger;


lib/easy_pool.js
var pool_module = require('generic-pool');
var mysql = require('mysql');

var conn_method = 'createConnection';
if (mysql.hasOwnProperty('createClient')) {
    conn_method = 'createClient';
}

/** @ignore */
var pool = {
    /** @ignore */
    create: function (settings) {
        return pool_module.Pool({
            name: 'mysql_' + settings.database,

            create: function (callback) {
                var client;
                if (mysql.hasOwnProperty(conn_method)) {
                    client = mysql[conn_method](settings);
                } else {
                    console.log("\nEasyMySQL [WARNING]: node-mysql 0.9.1 support is deprecated.\n");
                    client = new mysql.Client(settings);
                    client.connect();
                }
                callback(null, client);
            },

            destroy: function (client) {
                client.end();
            },

            max: settings.pool_size || 10,

            idleTimeoutMillis: settings.idle_timeout || 5000,

            log: settings.log || false
        });
    }
};

/** @namespace */
var easy_pool = (function () {
    var pool_cache = {};

    /**
     * @function
     * @param {object} settings - A settings object.
     * Settings properties:
     * <br><pre>
     *   user         : (required) - MySQL database user.
     *   database     : (required) - MySQL database to connect to.
     *   password     : (optional) - default: null
     *   host         : (optional) - default: localhost
     *   port         : (optional) - default: 3306
     *   pool_size    : (optional) - default: 10
     *   idle_timeout : (optional) - Timeout in milliseconds. default: 5000
     *   log          : (optional) - If true, log to console. If a function is passed in,
     *                               use that function for logging.
     *                               default: false.
     * </pre>
     *
     * @returns {object} a node-pool (generic-pool) Pool object.
     * @see <a href='https://github.com/coopernurse/node-pool'>node-pool</a>.
     */
    function fetch(settings) {
        var cache_key = JSON.stringify(settings);
        if (!pool_cache[cache_key]) {
            pool_cache[cache_key] = pool.create(settings);
        }
        return pool_cache[cache_key];
    }

    return {
        fetch: fetch
    };
}());

module.exports = easy_pool;


lib/easy_client.js
var mysql     = require('mysql');
var easy_pool = require('./easy_pool');

/**
 * @class
 * @property {object} client - A MySQL Client object.
 * @property {object} pool - The connection pool to use (if set).
 */
function EasyClient(client, pool) {
    this.client = client;
    this.pool = pool;
}

/**
 * This just delegates to the client query.
 * @param {string} sql - The sql query to execute.
 * @param {array} query_params (optional) - The query params to pass in to the query.
 * @param {function} cb - callback function
 * @returns {object} - The result returned by node-mysql.
 */
EasyClient.prototype.query = function (sql, query_params, cb) {
    this.client.query(sql, query_params, cb);
};

/**
 * If we're using a pool, release the client.
 * If we're using a client directly, call its end() function
 */
EasyClient.prototype.end = function () {
    if (this.pool) {
        this.pool.release(this.client);
    } else {
        this.client.end();
    }
};

/**
 * @private
 */
function get_client_from_pool(pool, cb) {
    pool.acquire(function (err, client) {
        if (err) {
            cb(err, null);
        } else if (!client) {
            // What should we do here?  Maybe throw the error?
            cb(new Error("Client not acquired"), null);
        } else {
            cb(null, new EasyClient(client, pool));
        }
    });
}

/**
 * Create an instance of EasyClient.
 * @param {object} settings - same settings as EasyMySQL.connect
 * @param {function} cb - callback function
 * @see EasyMySQL.connect
 */
EasyClient.fetch = function (settings, cb) {
    if (settings.pool) {
        get_client_from_pool(settings.pool, cb);
    } else if (settings.use_easy_pool) {
        var pool = easy_pool.fetch(settings);
        get_client_from_pool(pool, cb);
    } else {
        var client;
        if (mysql.hasOwnProperty('createConnection')) {
            client = mysql.createConnection(settings);
        } else if (mysql.hasOwnProperty('createClient')) {
            client = mysql.createClient(settings);
        } else {
            console.log("\nEasyMySQL [WARNING]: node-mysql 0.9.1 support is deprecated.\n");
            client = new mysql.Client(settings);
            client.connect();
        }
        cb(null, new EasyClient(client));
    }
};

exports.fetch = EasyClient.fetch;


lib/easy_mysql.js
var util = require('util');
var EasyClient = require('./easy_client');

/**
 * @class
 *
 * @description Note: use connect(), connect_with_easy_pool(),
 * and connect_with_pool() to create EasyMySQL objects.
 * This hides the implementation details and will make future
 * refactorings easier if the underlying implementation changes
 * (e.g., we decide not to use a class).
 *
 * @property {EasyClient} client - The EasyClient object to use.
 * @property {object} settings - The settings object passed into the constructor.
 * @see EasyMySQL.connect
 * @see EasyMySQL.connect_with_easy_pool
 * @see EasyMySQL.connect_with_pool
 */
function EasyMySQL(settings) {
    if (!(this instanceof EasyMySQL)) {
        return new EasyMySQL(settings);
    }
    var self = this;
    self.client = null;
    self.settings = settings;
    self.logging = settings.logging;

    /**
     * @private
     */
    this.fetch_client = function (cb) {
        EasyClient.fetch(self.settings, function (err, client) {
            if (err) {
                cb(err);
            } else {
                self.client = client;
                cb(null, client);
            }
        });
    };

    /**
     * @private
     */
    this.log_error = function (err, sql, query_params) {
        if (self.logging && self.logging.events && self.logging.events.error && self.logging.events.error.level) {
            var level = self.logging.events.error.level;

            var msg = 'EasyMySQL Error: ' + util.inspect(err) +
                ' - SQL: ' + sql + ' - Params: ' + util.inspect(query_params);
            if (err.stack) {
                msg += " - Stack: " + err.stack;
            }
            self.logging.logger[level](msg);
        }
    };
}

/**
 * Execute arbitrary SQL query.
 *
 * @param {string} sql - The sql query to execute.
 * @param {array} query_params (optional) - The query params to pass in to the query.
 * @param {function} cb - callback function
 * @returns {object} - The result returned by node-mysql.
 *
 * @example
 *    var sql = "insert into users(login) values(?)";
 *    mysql_easy.execute(sql, ['bob'], function(err, result) { });
 */
EasyMySQL.prototype.execute = function (sql, query_params, cb) {
    var self = this;
    if (typeof query_params === 'function') {
        cb = query_params;
        query_params = [];
    }
    self.fetch_client(function (err, easy_client) {
        if (err) {
            cb(err);
            if (easy_client) {
                easy_client.end();
            }
            return;
        } else {
            easy_client.query(sql, query_params, function (err, results) {
                if (err) {
                    self.log_error(err, sql, query_params);

                    if (easy_client) {
                        easy_client.end();
                    }

                    return cb(err, null);
                } else {
                    cb(null, results);
                    easy_client.end();
                }
            });
        }
    });
};

/**
 * Get a single result from SQL query.
 *
 * Note:
 * Be sure to add a limit statement or a 'where' clause that limits the results to one row.
 * This function only returns one result, but it doesn't add a limit statement for you.
 *
 * @param {string} sql - The sql query to execute.
 * @param {array} query_params (optional) - The query params to pass in to the query.
 * @param {function} cb - callback function
 *
 * @returns {object} - The first result of the query. Returns null if no results found.
 *
 * @example
 *    var sql = "select id, login from users where id = ?";
 *    mysql_easy.get_one(sql, [123], function(err, result) {
 *      console.log(result);
 *    });
 *
 *    >> {id:123, login:"jim@example.com"}
 */
EasyMySQL.prototype.get_one = function (sql, query_params, cb) {
    var self = this;
    if (typeof query_params === 'function') {
        cb = query_params;
        query_params = [];
    }
    self.execute(sql, query_params, function (err, results) {
        if (err) {
            cb(err, null);
        } else if (results.length === 0) {
            cb(null, null);
        } else {
            cb(null, results[0]);
        }
    });
};

/**
 * Get array of results from SQL query.
 *
 * @param {string} sql - The sql query to execute.
 * @param {array} query_params (optional) - The query params to pass in to the query.
 * @param {function} cb - callback function
 *
 * @returns {array} Array of the results of the query.  Returns empty array if no results found.
 *
 * @example
 *    var sql = "select id, login from users limit 2";
 *    mysql_easy.get_all(sql, function(err, result) {
 *      console.log(result);
 *    });
 *
 *    >> [{id:123, login:"jim@example.com"}, {id:456, login:"bob@example.com"}]
 */
EasyMySQL.prototype.get_all = function (sql, query_params, cb) {
    var self = this;
    if (typeof query_params === 'function') {
        cb = query_params;
        query_params = [];
    }
    self.execute(sql, query_params, function (err, results) {
        if (err) {
            cb(err, null);
        } else if (results.length === 0) {
            cb(null, []);
        } else {
            cb(null, results);
        }
    });
};

/**
 * Alias for execute
 * @see EasyMySQL#execute
 */
EasyMySQL.prototype.query = EasyMySQL.prototype.execute;

/*
 * Alias for get_one
 * @see EasyMySQL#get_one
 */
EasyMySQL.prototype.getOne = EasyMySQL.prototype.get_one;

/**
 * Alias for get_one
 * @see EasyMySQL#get_one
 */
EasyMySQL.prototype.one = EasyMySQL.prototype.get_one;

/**
 * Alias for get_all
 * @see EasyMySQL#get_all
 */
EasyMySQL.prototype.getAll = EasyMySQL.prototype.get_all;

/**
 * Alias for get_all
 * @see EasyMySQL#get_all
 */
EasyMySQL.prototype.all = EasyMySQL.prototype.get_all;

/**
 * Directly establish a single connection for each query.
 * This is probably not a good idea for production code, but may be fine for code
 * where you don't want to set up a pool, such as in unit tests.
 *
 * @param {object} settings - An object of MySQL settings. Settings properties:
 * <br><pre>
 *   user          : (required) - Database user.
 *   database      : (required) - Database to connect to.
 *   password      : (optional) - default: null
 *   host          : (optional) - default: localhost
 *   port          : (optional) - default: 3306
 * </pre>
 *
 * @returns {EasyMySQL} - returns an EasyMySQL object.
 *
 * @example
 * var settings = {
 *     user     : 'myuser',
 *     password : 'mypass',
 *     database : 'mydb'
 * };
 *
 * var easy_mysql = EasyMySQL.connect(settings);
 * easy_mysql.get_one("select * from foo where bar = ?", ['jazz'], function (err, result) {
 *     // do stuff
 * });
 */
EasyMySQL.connect = function (settings) {
    return new EasyMySQL(settings);
};

/**
 * Use built-in easy_pool.js, which uses generic-pool.
 *
 * @see <a href='https://github.com/coopernurse/node-pool'>node-pool</a>.
 * @see easy_pool
 *
 * @param {object} settings - An object of MySQL settings. Settings properties:
 * <br><pre>
 *   user          : (required) - Database user.
 *   database      : (required) - Database to connect to.
 *   password      : (optional) - default: null
 *   host          : (optional) - default: localhost
 *   port          : (optional) - default: 3306
 *   pool_size     : (optional) - The size of the pool in easy_pool.js. Default: 10
 * </pre>
 *
 * @returns {EasyMySQL} - returns an EasyMySQL object.
 *
 * @example
 * var settings = {
 *     user      : 'myuser',
 *     password  : 'mypass',
 *     database  : 'mydb',
 *     pool_size : 50
 * };
 *
 * var easy_mysql = EasyMySQL.connect_with_easy_pool(settings);
 */
EasyMySQL.connect_with_easy_pool = function (settings) {
    settings.use_easy_pool = true;
    return new EasyMySQL(settings);
};

/**
 * Use a custom pool:
 *
 * @see easy_pool
 *
 * @param {object} pool - A connection pool. It must have functions named 'acquire' and 'release'.
 * See example in easy_pool.js.
 *
 * @returns {EasyMySQL} - returns an EasyMySQL object.
 *
 * @example
 * var my_pool = //.. create your own pool however you want.
 *
 * var easy_mysql = EasyMySQL.connect_with_pool(pool);
 */
EasyMySQL.connect_with_pool = function (pool, settings) {
    var args = {};
    if (settings) {
        args = settings;
    }
    args.pool = pool;
    return new EasyMySQL(args);
};

exports.connect = EasyMySQL.connect;
exports.connect_with_easy_pool = EasyMySQL.connect_with_easy_pool;
exports.connect_with_pool = EasyMySQL.connect_with_pool;
